sgkit/distance/core.py

```diff
@@ -0,0 +1,87 @@
+import dask.array as da
+import numpy as np
+
+from sgkit.typing import ArrayLike
+
+
+def pairwise(
+    x: ArrayLike,
+    metric_func: np.ufunc,
+) -> ArrayLike:
+    """A generic pairwise function for any separable distance metric.
+    This calculates the pairwise distance between a set of vectors in the
+    given two-dimensional array, using the ``metric_func`` ufunc. To illustrate
+    the algorithm consider the following (4, 5) two dimensional array:
+
+    [e.00, e.01, e.02, e.03, e.04]
+    [e.10, e.11, e.12, e.13, e.14]
+    [e.20, e.21, e.22, e.23, e.24]
+    [e.30, e.31, e.32, e.33, e.34]
+
+    The rows of the above matrix are the set of vectors. Now lets label all
+    the vectors as v0, v1, v2, v3
+
+    The result will be a two dimensional symmetric matrix which will contain
+    the distance between all pairs. Since there are 4 vectors, calculating the
+    distance of each vector with every other vector, will result in 16
+    distances and the resultant array will be of size (4, 4) as follows:
+
+    [v0.v0, v0.v1, v0.v2, v0.v3]
+    [v1.v0, v1.v1, v1.v2, v1.v3]
+    [v2.v0, v2.v1, v2.v2, v2.v3]
+    [v3.v0, v3.v1, v3.v2, v3.v3]
+
+    The (i, j) position in the resulting array (matrix) denotes the distance
+    between vi and vj vectors.
+
+    The given array can be either of type numpy or dask array. For better
```

Comment:
I think you can remove this paragraph.  The input types should be clear from the parameter docs, it's not clear what "suitable chunking" is, and I'm not sure that parallelizing with dask would necessarily increase performance.  I would assume that's not true for the majority of smaller inputs.